### PR TITLE
chore(config): Set up `ConfigurableRef` for schema functions

### PR DESCRIPTION
**Pull request by bruceg**
The config schema functions currently dispatch which configurable type to use
based on a type parameter to the functions. In order to make this dispatch
dynamic, a new `struct ConfigurableRef` is set up that contains references to
the functions for a given type implementing `trait Configurable`. This is then
used in the schema functions to dispatch their operation, eliminating the type
bound. This will be used by a follow-up change to allow for dynamically
registrable transforms.

### CODE REVIEW

File: lib/vector-config/src/configurable.rs [hunk 1]
```diff
@@ -65,6 +65,14 @@ where
     /// If an error occurs while generating the schema, an error variant will be returned describing
     /// the issue.
     fn generate_schema(gen: &RefCell<SchemaGenerator>) -> Result<SchemaObject, GenerateError>;
+
+    /// Create a new configurable reference table.
+    fn make_ref() -> ConfigurableRef
```
Comment (bruceg) [inline lib/vector-config/src/configurable.rs:70]: Note: I'm not particularly happy about this function name, so if you have a better idea I would welcome hearing it.

Comment (tobz) [reply lib/vector-config/src/configurable.rs:70]: I'd probably just call it `as_configurable_ref`, which feels a bit more Rust-y, even if the return value isn't _literally_ a reference.

File: lib/vector-config/src/schema.rs [hunk 2]
```diff
@@ -232,12 +231,12 @@ where
...
-    T: Configurable,
-{
-    // Generate the actual schema for the element type `T`.
-    let element_schema = get_or_generate_schema::<T>(gen, None)?;
+pub(crate) fn generate_set_schema(
+    config: &ConfigurableRef,
+    gen: &RefCell<SchemaGenerator>,
+) -> Result<SchemaObject, GenerateError> {
+    // Generate the actual schema for the element type.
+    let element_schema = get_or_generate_schema(config, gen, None)?;
```
Comment (tobz) [inline lib/vector-config/src/schema.rs:239]: Why are we making this function, and some others, only public in the crate itself? Clippy lints?

File: lib/vector-config/src/schema.rs [hunk 3]
```diff
@@ -6,18 +6,20 @@ use serde_json::{Map, Value};
...
 };
 
+pub fn apply_base_metadata(schema: &mut SchemaObject, metadata: Metadata) {
+    apply_metadata(&<()>::make_ref(), schema, metadata)
+}
+
 /// Applies metadata to the given schema.
 ///
 /// Metadata can include semantic information (title, description, etc), validation (min/max, allowable
 /// patterns, etc), as well as actual arbitrary key/value data.
```
Comment (tobz) [inline lib/vector-config/src/schema.rs:19]: Should move this doc comment to `apply_base_metadata`.

File: lib/vector-config/src/schema.rs [hunk 4]
```diff
@@ -479,30 +461,32 @@ pub fn generate_internal_tagged_variant_schema(
...
+    let (mut schema, metadata) = match &config.referenceable_name() {
         // When `T` has a referenceable name, try looking it up in the schema generator's definition
         // list, and if it exists, create a schema reference to it. Otherwise, generate it and
         // backfill it in the schema generator.
         Some(name) => {
-            if !gen.borrow().definitions().contains_key(name) {
+            // Creating a string just to borrow it is ugly, but that's required by the types in the
+            // schema generator definitions.
+            if !gen.borrow().definitions().contains_key(&name.to_string()) {
                 // In order to avoid infinite recursion, we copy the approach that `schemars` takes and
```
Comment (tobz) [inline lib/vector-config/src/schema.rs:490]: `name` should be `&'static str`, and `IndexMap::contains_key` should function equivalently to `HashMap::contains_key`, where if `K` is `String`, we should be able to use `&str` as a key for lookups.. so I feel like I'm missing something here. 🤔

**Review by tobz — COMMENTED** [review]
This generally looks good to me, but I had some questions/thoughts.

File: lib/vector-config/src/schema.rs (hunk 2 above)
Comment (bruceg) [reply lib/vector-config/src/schema.rs:239]: No particular reason, other than going through and seeing which are actually used outside of the crate.

File: lib/vector-config/src/schema.rs (hunk 4 above)
Comment (bruceg) [reply lib/vector-config/src/schema.rs:490]: This is the error I get just using `name` here. I couldn't figure another formulation that works.
```
error[E0277]: the trait bound `std::string::String: Borrow<&str>` is not satisfied
   --> lib/vector-config/src/schema.rs:487:57
    |
487 |             if !gen.borrow().definitions().contains_key(name) {
    |                                            ------------ ^^^^ the trait `Borrow<&str>` is not implemented for `std::string::String`
    |                                            |
    |                                            required by a bound introduced by this call
    |
    = help: the trait `Borrow<str>` is implemented for `std::string::String`
    = note: required for `&str` to implement `Equivalent<std::string::String>`
note: required by a bound in `IndexMap::<K, V, S>::contains_key`
   --> /home/bruce.guenter/.cargo/registry/src/github.com-1ecc6299db9ec823/indexmap-1.9.2/src/map.rs:400:19
    |
400 |         Q: Hash + Equivalent<K>,
    |                   ^^^^^^^^^^^^^ required by this bound in `IndexMap::<K, V, S>::contains_key`

For more information about this error, try `rustc --explain E0277`.
```

Comment (tobz) [reply lib/vector-config/src/schema.rs:490]: Ohhh. Drop that extra reference on line 482.

File: lib/vector-config/src/schema.rs (hunk 3 above)
Comment (bruceg) [reply lib/vector-config/src/schema.rs:15]: Fixed in f27b31f2b

File: lib/vector-config/src/schema.rs (hunk 4 above)
Comment (bruceg) [reply lib/vector-config/src/schema.rs]: Oh, duh, missed the obvious there. Fixed in eddc6a19f

File: lib/vector-config/src/schema.rs (hunk 2 above)
Comment (tobz) [reply lib/vector-config/src/schema.rs:239]: Works for me. 🤷🏻

**Review by tobz — APPROVED** [review]
Looks good to me. 👍🏻

**Pull request merged**